aten/src/ATen/MemoryOverlap.cpp

```diff
@@ -34,4 +34,39 @@ void assert_no_internal_overlap(TensorImpl* t) {
     "performing the operation.");
 }
 
+MemOverlapStatus get_overlap_status(const Tensor& a, const Tensor& b) {
+  return get_overlap_status(a.unsafeGetTensorImpl(), b.unsafeGetTensorImpl());
+}
+
+MemOverlapStatus get_overlap_status(TensorImpl* a, TensorImpl* b) {
+  if (!a->is_contiguous() || !b->is_contiguous()) {
+    return MemOverlapStatus::TOO_HARD;
+  }
+  if (a->storage().data() == b->storage().data()) {
+    const auto a_begin = static_cast<char*>(a->data());
+    const auto a_end = a_begin + a->numel() * a->itemsize();
+    const auto b_begin = static_cast<char*>(b->data());
+    const auto b_end = b_begin + b->numel() * b->itemsize();
+
+    if (a_begin == b_begin && a_end == b_end) {
+      return MemOverlapStatus::FULL;
+    }
+    if (a_begin < b_end && b_begin < a_end) {
```

Comment:
Not going to work when two tensors comes one after another in the same storage.

Reply:
If b_begin < a_end, doesn't that mean the two tensors can't come directly after another in the same storage?

Reply:
a_end - points to the memory After tensor, not to the last byte of the tensor. Basically tensor allocated at `[a_begin, a_end)` in math notation.

Reply:
Sorry, I think I'm missing something.

Let's say that `b` is directly after `a` in memory.

`a_end` points to the memory right after `a`. so `a_end` is the same thing as `b_begin`. `b_begin < a_end` would return false, so there is no partial memory overlap, which I think is what we wanted here.

Reply:
Right, my bad.